src/bridge/crypto/ec.cc

```diff
@@ -0,0 +1,303 @@
+    int ec_size = ECDSA_size(ec_key);
+    if (ec_size < 0) {
+      KMSENGINE_SIGNAL_ERROR(Status(StatusCode::kFailedPrecondition,
+                                    "ECDSA_size(ec_key) was < 0"));
+      return 0;
+    }
+    if (signature.length() > static_cast<std::string::size_type>(ec_size)) {
+      KMSENGINE_SIGNAL_ERROR(
+          Status(StatusCode::kFailedPrecondition,
+                 "Generated signature length was unexpectedly larger than "
+                 "ECDSA_size(ec_key)"));
+      return 0;
+    }
+    signature.copy(reinterpret_cast<char *>(signature_return),
+                   signature.length());
+  }
+  return 1;
+}
+
+// Engine implementation for `ECDSA_sign_setup`.
+//
+// Used to precompute parts of the signing operation. `ec_key` is the private
+// EC key and `context` is a pointer to `BN_CTX` structure (meaning "BIGNUM
+// context"). `context` may be null. The precomputed values returned in `kinv`
+// and `rp` are intended to be used in a later call to `ECDSA_sign_ex` or
+// `ECDSA_do_sign_ex`.
+//
+// Since the engine is not actually performing any crypto operations, we don't
+// really need this function. Our implementation ignores `ec_key` and `context`
+// and sets `*kinv` and `*rp` to nullptr.
+//
+// Returns 1 on success; otherwise, returns 0.
+int SignSetup(EC_KEY */*ec_key*/, BN_CTX */*context*/, BIGNUM **kinv,
+              BIGNUM **rp) {
+  // Set `*kinv` and `*rp` to nullptr, since calling code should not expect
+  // `*kinv` and `*rp` to be populated after calling this function.
+  *kinv = nullptr;
+  *rp = nullptr;
+
+  return 1;
+}
+
+// Engine implementation for `ECDSA_do_sign_ex`.
+//
+// Computes a digital signature of the dgst_len bytes hash value dgst using the
+// private key eckey and the optional pre-computed values kinv and rp. The
+// signature is returned in a newly allocated ECDSA_SIG structure (or NULL on
+// error).
+//
+// Returns a pointer to an allocated ECDSA_SIG structure or NULL on error.
+ECDSA_SIG *DoSignEx(const unsigned char *digest_bytes, int digest_length,
+                    const BIGNUM *kinv, const BIGNUM *rp,
+                    EC_KEY *ec_key) {
+  OpenSslEcdsaSignature ecdsa_signature = MakeEcdsaSignature();
+  if (ecdsa_signature == nullptr) {
+    KMSENGINE_SIGNAL_ERROR(
+        Status(StatusCode::kResourceExhausted, "No memory available"));
+    return nullptr;
+  }
+
+  // We reuse the engine's implementation for `ECDSA_sign_ex` to compute the
+  // signature here.
+  std::string signature('\0', ECDSA_size(ec_key));
+  unsigned int signature_length;
+  if (!ECDSA_sign_ex(0, digest_bytes, digest_length,
+                     reinterpret_cast<unsigned char *>(&signature[0]),
+                     &signature_length, kinv, rp, ec_key)) {
+    KMSENGINE_SIGNAL_ERROR(
+        Status(StatusCode::kInternal, "ECDSA_sign_ex failed"));
+    return nullptr;
+  }
+
+  // BN_bin2bn(&signature[0], signature_length, nullptr);
```

Comment:
Why is this line commented out?

Reply:
This function actually looks like it's incomplete (sorry!). It's supposed to return an instance of `ECDSA_SIG`, which separates the `r` and `s` components of an EC signature into two `BN` (bignums). I think I was not immediately able to figure out how to convert the DER-encoded signature returned by Cloud KMS API into an `ECDSA_SIG`. However, I think [`d2i_ECDSA_SIG`](https://www.openssl.org/docs/man1.0.2/man3/d2i_ECDSA_SIG.html) is the right place to look as to how to convert the `unsigned char *` signature from `SignEx` to an `ECDSA_SIG` (though I wasn't able to immediately figure it out).